### PR TITLE
Adding Handled/Unhandled payload properties

### PR DESCRIPTION
**Pull request by Cawllec**
- Adds `defaultSeverity`, `unhandled`, and `severityReasons` payload properties to support upcoming handled/unhandled functionality
- Above properties added through **options parameters
- System error handler, log level handlers, and middleware error handlers set up with severity reasons
_references [PLAT_209](https://bugsnag.atlassian.net/browse/PLAT-209)_

### CODE REVIEW

File: bugsnag/wsgi/middleware.py [hunk 1]
```diff
@@ -28,13 +28,23 @@ class WrappedWSGIApp(object):
 
     def __init__(self, application, environ, start_response):
         self.environ = environ
+        self.severity_reason = {
```
Comment (snmaynard) [inline bugsnag/wsgi/middleware.py]: This should be a constant rather than a member on the middleware

File: bugsnag/client.py [hunk 2]
```diff
@@ -117,12 +120,24 @@ def deliver(self, notification):  # type: (Notification) -> None
...
+        initial_unhandled = notification.unhandled
+        initial_severity_reason = notification.severity_reason
+
         def send_payload():
             if notification.api_key is None:
                 bugsnag.logger.warning(
                     "No API key configured, couldn't notify")
                 return
 
+            notification.default_severity = initial_default_severity
```
Comment (snmaynard) [inline bugsnag/client.py]: I dont think we need to reset any changes to these variables in here. If people change them then they are in expert mode and thats fine.

File: bugsnag/celery/__init__.py [hunk 3]
```diff
@@ -14,7 +14,14 @@ def failure_handler(sender, task_id, exception, args, kwargs, traceback, einfo,
 
     bugsnag.auto_notify(exception, traceback=traceback,
                         context=sender.name,
-                        extra_data=task)
+                        extra_data=task,
+                        unhandled=True,
```
Comment (snmaynard) [inline bugsnag/celery/__init__.py]: is this not implied because auto notify?

File: bugsnag/legacy.py [hunk 4]
```diff
@@ -47,6 +47,9 @@ def notify(exception, **options):
     """
     Notify bugsnag of an exception.
     """
+    if 'severity' in options:
```
Comment (snmaynard) [inline bugsnag/legacy.py]: this will trigger for all auto notify calls which is incorrect, you still want defaultSeverity to be true in those cases

**Comment by Cawllec** [comment]
I've updated the notify default-severity check, that shouldn't automatically set default_severity to false when unhandled calls to it are made.  I've taken out the overriding of severity_reason and unhandled, but this is a pattern in some of the other notifiers (whether through making them protected properties or by overriding them if necessary) which may need discussion

File: bugsnag/client.py [hunk 5]
```diff
@@ -117,12 +122,22 @@ def deliver(self, notification):  # type: (Notification) -> None
...
                 bugsnag.logger.warning(
                     "No API key configured, couldn't notify")
                 return
 
+            if initial_severity != notification.severity:
+                notification.severity_reason = {
+                    'type': 'userCallbackSetSeverity'
+                }
+            else:
+                notification.severity_reason = initial_reason
```
Comment (snmaynard) [inline bugsnag/client.py]: Why do we do this here?

File: bugsnag/client.py [hunk 6]
```diff
@@ -150,6 +165,8 @@ class ClientContext(object):
     def __init__(self, client, exception_types=None, **options):
         self.client = client
         self.options = options
+        if 'severity' in options:
+            options['severity_reason'] = dict(type='contextSpecifiedSeverity')
```
Comment (snmaynard) [inline bugsnag/client.py]: Should this be `userContextSetSeverity`?

File: bugsnag/handlers.py [hunk 7]
```diff
@@ -37,7 +37,16 @@ def emit(self, record):
             if path in record.pathname:
                 return
 
-        options = {'meta_data': {}}
+        options = {
+            'meta_data': {},
+            'unhandled': False,
+            'severity_reason': {
+                'type': 'logs',
```
Comment (snmaynard) [inline bugsnag/handlers.py]: This should be log

File: bugsnag/client.py (hunk 5 above)
Comment (Cawllec) [reply bugsnag/client.py]: This enforces that users cannot overwrite the severity_reason in callbacks or middleware.  I think we previously said it wouldn't matter if someone chose to do so, but we've added it as one of our testing parameters and are enforcing the same thing across notifiers.

File: bugsnag/client.py (hunk 6 above)
Comment (Cawllec) [reply bugsnag/client.py]: This one seems to already have been fixed

**Pull request merged**